date-time/src/commonMain/kotlin/com/splendo/kaluga/datetime/timer/Timer.kt

```diff
@@ -0,0 +1,70 @@
+/*
+ Copyright 2021 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+package com.splendo.kaluga.datetime.timer
+
+import com.splendo.kaluga.base.utils.firstInstance
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.transformWhile
+import kotlin.time.Duration
+
+/** A timer ticking a certain [duration]. */
+interface Timer {
```

Comment:
you can test this with `sleep` on the main dispatcher, but if you make `val mark = TimeSource.Monotonic.markNow()` a default argument (through the chain) instead of a hard coding it in you can throw in a Fake(?) TimeSource.

Reply:
I'm not sure I understood the idea, but it looks to me as if `Timer:StateFlow` would be leaking implementation details in this case.

as for the `StateFlowRepo` the name suggests a `contains` relation to the `StateFlow`. IMO we'd have to rename it to something like `XxxxStateFlow` to make it `is a` relation

Reply:
What implementation detail is leaked?

Reply:
oohh. I DID misunderstand the idea. I took another look, so instead of 
```
    private val stateRepo = TimerStateRepo(duration, interval, timeProvider, coroutineScope)
    override val state: StateFlow<Timer.State> = stateRepo.stateFlow
```
there would be
```
    override val state: StateFlow<Timer.State> = TimerStateRepo(duration, interval, timeProvider, coroutineScope)
```

yeah, it's nice, but I still stand by the `is-a` vs `contains` comment. if a repo would implement a `StateFlow`, it will need a different name.

Reply:
`TimerStateRepo` is already a `Flow`, this would just change it to `StateFlow`.

This is the reason I am suggesting a change, it's weird there it both `is-a` Flow and `has-a` StateFlow.

Reply:
IMO, a `Repo` suffix suggests that it contains something, so 
```
interface StateFlowRepo<S : State> {
    val stateFlow: StateFlow<S>
}
```
makes a perfect sense to me, while 
```
class StateRepo ... :  Flow<S>
```
is somewhat questionable. personally I'd prefer 
```
class StateFlow ... :  Flow<S> // StateRepo renamed. clear `is-a` 
//or 
class StateRepo ... : StateFlowRepo{ // not implementing Flow any more, clear `contains`
   val stateFlow: StateFlow<S>
}

Reply:
The question was for getting feedback however, in particular if you are interested in `Timer:StateFlow<Timer.State<`. 

The name `StateFlowRepo` IMHO already expresses it it both a `StateFlow` and a `Repo` (if that's a bad idea that it is both the moment to change that has long past now), but then for some reason it's not actually a `StateFlow` but just `Flow`. Fixing this remains to be done, the reason to mention it here was in case you need it.

My main reasoning is it would encourage using `StateFlow<Timer.State>` e.g. when injecting it somewhere that just needs a view of the time. Of course this can still be done, it's just less encouraged because it's in a field. 

But the answer is clear, you don't want your `Timer` to be a (State)Flow.

I would consider then instead splitting the interface in `Timer` and `ControllableTimer` (or better names) that adds the `start/stop` to encourage places that only want a view of the Timer to need something that controls it.